openmmtools/cache.py

```diff
@@ -0,0 +1,352 @@
+#!/usr/bin/env python
+
+# =============================================================================
+# MODULE DOCSTRING
+# =============================================================================
+
+"""
+Provide cache classes to handle creation of OpenMM Context objects.
+
+"""
+
+
+# =============================================================================
+# GLOBAL IMPORTS
+# =============================================================================
+
+import copy
+import collections
+
+from simtk import openmm
+
+from openmmtools import utils
+
+
+# =============================================================================
+# GENERAL LRU CACHE
+# =============================================================================
+
+class LRUCache(object):
+    """A simple LRU cache.
+
+    It can be configured to have a maximum number of elements (capacity)
+    and an element expiration (time_to_live) measured in number of accesses
+    to the cache. Both read and write operations count as an access, but
+    only successful reads (i.e. those not raising KeyError) increment the
+    counter.
+
+    Parameters
+    ----------
+    capacity : int, optional
+        Maximum number of elements in the cache. When set to None, the
+        cache has infinite capacity (default is None).
+    time_to_live : int, optional
```

Comment:
I think we can always add other behaviors later if we find they will be useful. I was originally wondering if it would be useful to have a few optional `time_to_live` arguments, but I think this single argument that could later have multiple interpretations based on the units attached to the value passed is just fine.